validity/src/utils.rs

```diff
@@ -33,6 +37,93 @@ pub fn find_gaps(overall_start: i64, overall_end: i64, ranges: &[(i64, i64)]) ->
     gaps
 }
 
+/// Determines the block ranges to be proven based on a set of ranges and a gas limit.
+///
+/// Given a set of block ranges that overlap at most on the boundaries and a gas limit, this
+/// function calculates and returns the specific block ranges that need to be proven based on
+/// the gas used in each block. Ensures that all disjoint ranges are fully covered by creating
+/// ranges that don't exceed the gas limit.
+///
+/// # Arguments
+/// * `disjoint_ranges` - The ranges to split based on gas usage
+/// * `gas_limit` - Maximum gas allowed per range
+/// * `block_infos` - Map of block number to BlockInfo for all blocks in the ranges
+///
+/// # Example
+///
+/// ```ignore
+/// use op_succinct_validity::get_ranges_to_prove_by_gas;
+/// use std::collections::HashMap;
+///
+/// let disjoint_ranges = [(0, 4)];
+/// let gas_limit = 100_000_000; // 100 million gas
+/// let mut block_infos = HashMap::new();
+/// // Add BlockInfo for blocks 1-4...
+///
+/// // If blocks 1-4 use 30M gas each:
+/// let ranges_to_prove = get_ranges_to_prove_by_gas(&disjoint_ranges, gas_limit, &block_infos)?;
+/// // Result: [(0, 3), (3, 4)] - first 3 blocks (90M gas), then last block (30M gas)
+/// ```
+pub fn get_ranges_to_prove_by_gas(
+    disjoint_ranges: &[(i64, i64)],
+    evm_gas_limit: u64,
+    range_proof_interval: i64,
+    block_infos: &HashMap<i64, BlockInfo>,
+) -> Result<Vec<Range<i64>>> {
+    let mut ranges = Vec::new();
+
+    for &(start, end) in disjoint_ranges {
+        let mut current_start = start;
+        let mut accumulated_gas = 0u64;
+
+        // Process blocks in the range (start+1 to end inclusive)
+        // Note: We don't prove the start block, only use its hash as starting point
+        for block_num in (start + 1)..=end {
```

Comment:
If start block is 0 and end block is 1, neither of the gas limit or block limit condition will be checked. Can you make a fix and add test case for it?

Reply:
I added [a test](https://github.com/succinctlabs/op-succinct/pull/576/files#diff-ee536376aa679baaca4d2f6646dcc842c97ac6d8cae7da1cca03936cc9d0bedaR411) and it's passing.

I think the range [is added here](https://github.com/succinctlabs/op-succinct/pull/576/files#diff-ee536376aa679baaca4d2f6646dcc842c97ac6d8cae7da1cca03936cc9d0bedaR113).